### ISSUE TITLE
Add Stunnel

### ISSUE THREAD

**Issue by joejoe2017** [pivpn/pivpn#525]
Hello, 

Any way you can provide me with the steps to add Stunnel (SSH) to Pivpn ?

**Comment by cezar97** [comment]
Hi,
you can follow this guide skipping the first steps to install redis https://www.digitalocean.com/community/tutorials/how-to-encrypt-traffic-to-redis-with-stunnel-on-ubuntu-16-04.

**Comment by joejoe2017** [comment]
Thank you for the reply, But this is really too complicated for me , do you have easier steps ?

**Comment by cezar97** [comment]
[BOTH ON CLIENT AND SERVER]
Update repos: `sudo apt update`
Install stunnel: `sudo apt install stunnel4`
Edit the following to start stunnel at boot: `sudo nano /etc/default/stunnel4` and change to `ENABLED=1`

[ON SERVER]
Create certificates with: `sudo openssl req -x509 -nodes -days 3650 -newkey rsa:2048 -keyout /etc/stunnel/openvpn-server.key -out /etc/stunnel/openvpn-server.crt`. For contry-name and other stuff asked you can simpy put anything you like.
Restrict key permissions: `sudo chmod 600 /etc/stunnel/openvpn-server.key`
Add a config file for stunnel: `sudo nano /etc/stunnel/openvpn.conf` and add the following:
```
pid = /run/stunnel-openvpn.pid

[openvpn-server]
cert = /etc/stunnel/openvpn-server.crt
key = /etc/stunnel/openvpn-server.key
accept = IP_OF_STUNNEL_MACHINE:DESIRED_PORT
connect = IP_OF_OPENVPN_SERVER:OPENVPN_PORT
```
Restart stunnel `sudo systemctl restart stunnel4.service`
Copy `/etc/stunnel/openvpn-server.crt` certificate to the client

[ON CLIENT]
Move the copied certificate in `/etc/stunnel/openvpn-server.crt`
Edit the config: `sudo nano /etc/stunnel/openvpn.conf` and add the following:
```
pid = /run/stunnel-openvpn.pid

[openvpn-client]
client = yes
accept = IP_OF_STUNNEL_CLIENT:DESIRED_PORT
connect = IP_OF_STUNNEL_SERVER:PORT_OF_STUNNEL_SERVER
CAfile = /etc/stunnel/openvpn-server.crt
verify = 4
```
Restart stunnel: `sudo systemctl restart stunnel4.service`

Note:
On server side, the accept statement contains the IP and the port stunnel will listen on. In the case of a VPS, the IP will be its public IP; in case of a machine behind NAT, you put in the local ip and configure port forwarding on the router. The connect statement contains the IP and the port of the OpenVPN server. In case you have the OpenVPN server on the same machine then as IP you write `127.0.0.1`.
On client side, the accept contains the IP and the port stunnel will listen on. The connect contains the IP of the stunnel server and its remote port.
Once everything is configured, to connect to OpenVPN inside stunnel's tunnel, you have to change the following line of the OVPN profile you use to connect so it looks like `remote IP_STUNNEL_CLIENT PORT_STUNNEL_CLIENT` instead of the previously `remote IP_OPENVPN_SERVER PORT_OPENVPN_SERVER` .
The openvpn client machine and the stunnel client machine shuold be different as running both stunnel and openvpn on the same pc wouldn't work due to routing issues.

PS:
> do you have easier steps ?

Sorry for this, I hadn't time for an explanation. :)

**Comment by joejoe2017** [comment]
Thank you for this clarification, though, I still have some question.
can this setup work on OPENVPN Router connecting to PIVPN(on Rasperrypi), as I can see you want me to edit the certificate on the Client. My client router or android are only using OPENVPN software that I import the OVPN file to it ?

**Comment by cezar97** [comment]
Stunnel server has 2 options to set: the accept indicates the ip and the port it accepts connections from stunnel client, the connect is the ip and the port where it redirects traffic coming from stunnel client.
The same with stunnel client: connects it's the ip and the port of stunnel server, where the ssl encrypted traffic goes, accept indicates the ip and the port other applications connects to.
In other words, all traffic sent to stunnel client, it is ssl encrypted and goes to stunnel server where it is redirected to the openvpn server, in this case. That's why in the ovpn profile you use to connect you have the ip and port of stunnel client instead of directly the openvpn machine.

**Comment by joejoe2017** [comment]
Thank you so much again about your reply, and excuse me for my null information. You cannot imagine how I appreciate your help
I will detailed my situation:

Server:
      Raspberry pi with PIVPN (Local IP : 10.8.20.131, open port for VPN is UDP:11945)
      Behind ASUS router
      DDNS : 123.asuscomm.com
  
     Can you please add the suitable number in the accept and connect fields
	pid = /run/stunnel-openvpn.pid
	[openvpn-server]
	cert = /etc/stunnel/openvpn-server.crt
	key = /etc/stunnel/openvpn-server.key
	accept = IP_OF_STUNNEL_MACHINE:DESIRED_PORT
	connect = IP_OF_OPENVPN_SERVER:OPENVPN_PORT

Client : GL-Int router with OPENWRT firmware
       Local IP: 192.168.1.2

	pid = /run/stunnel-openvpn.pid
         [openvpn-client]
         client = yes
        accept = IP_OF_STUNNEL_CLIENT:DESIRED_PORT
         connect = IP_OF_STUNNEL_SERVER:PORT_OF_STUNNEL_SERVER
         CAfile = /etc/stunnel/openvpn-server.crt
          verify = 4

**Comment by cezar97** [comment]
So, on your pi ( server ) install stunnel with the server configuration. The connect will be 127.0.0.1:11945 and the accept 10.8.20.131:443 ( as port you can put whatever you want but remember to forward that port on the router just like you did with openvpn ).
Now, on client side I'm not sure if it is possible to do that at the router level. Anyway, an example of a setup is the following. Install and configure stunnel as client on a pc, the accept will be, for example, 192.168.1.3:11945 and the connect ddns:443 . On the same network then instead of connecting directly to ddns:11945, you connect to 192.168.1.3:11945.
